File: scripts/verify-parser.mjs
// Checks lib/plan-parser.mjs against an independent reading of the same plans (scripts/verify-parser.py,
// Python's ElementTree), field by field. The analysis is only as good as the evidence the models receive, so
// every number the digest gives them must match the plan.
//
//   node scripts/verify-parser.mjs <folder or .sqlplan files...> [--verbose]
//
// Exit code 1 when any field disagrees.
import { execFileSync } from "node:child_process";
import { readFileSync, readdirSync, statSync } from "node:fs";
import { join } from "node:path";
import { parsePlanText } from "../lib/plan-parser.mjs";
import { decodeBytes } from "../lib/plan-decoder.mjs";

const args = process.argv.slice(2).filter((a) => !a.startsWith("--"));
const verbose = process.argv.includes("--verbose");
const walk = (p) =>
  statSync(p).isDirectory()
    ? readdirSync(p).flatMap((f) => walk(join(p, f)))
    : /\.(sqlplan|xml)$/i.test(p)
      ? [p]
      : [];
const files = args.flatMap(walk);

// Relative tolerance for floats the plan itself rounds (costs are printed to ~6 significant digits).
const close = (a, b, rel = 1e-4) =>
  a == null || b == null
    ? a == b
    : Math.abs(a - b) <= Math.max(1e-6, rel * Math.max(Math.abs(a), Math.abs(b)));
const norm = (s) => (s ?? "").replace(/\s+/g, " ").trim();

const problems = [];
let statementsChecked = 0,
  operatorsChecked = 0,
  fieldsChecked = 0;
const bad = (file, where, field, ours, truth) =>
  problems.push({ file, where, field, ours, truth });

for (const file of files) {
  const short = file.split("/").slice(-2).join("/");
  const truth = execFileSync("python3", ["scripts/verify-parser.py", file], {
    encoding: "utf8",
    maxBuffer: 512 * 1024 * 1024,
  })
    .trim()
    .split("\n")
    .filter(Boolean)
    .map((l) => JSON.parse(l));
  if (truth.some((t) => t.error)) {
    bad(short, "file", "xml", "parsed", truth[0].error);
    continue;
  }
  const ours = parsePlanText(decodeBytes(readFileSync(file)));
  // Match statements by their text (our ids are sequential, the plan's StatementId need not be), in order.
  const unused = [...ours];
  const take = (t) => {
    const key = norm(t.sql).slice(0, 120);
    const i = unused.findIndex(
      (d) =>
        norm(d.sql).slice(0, 120) === key ||
        (d.sqlTruncated && key.startsWith(norm(d.sql).slice(0, 60))),
    );
    return i === -1 ? null : unused.splice(i, 1)[0];
  };
  for (const t of truth) {
    const d = take(t);
    const where = `statement ${t.statementId}`;
    if (!d) {
      bad(short, where, "statement", "missing", t.type);
      continue;
    }
    statementsChecked++;
    const check = (field, a, b, ok) => {
      fieldsChecked++;
      if (!ok) bad(short, where, field, a, b);
    };
    // SQL text: the digest may truncate long statements and says so.
    const sql = norm(d.sql),
      tsql = norm(t.sql);
    check(
      "sql",
      sql.slice(0, 80),
      tsql.slice(0, 80),
      d.sqlTruncated ? tsql.startsWith(sql.replace(/…$/, "").trim().slice(0, 200)) : sql === tsql,
    );
    check("subtreeCost", d.subtreeCost, t.cost, close(d.subtreeCost, t.cost));
    check("operatorCount", d.operatorCount, t.operators.length, d.operatorCount === t.operators.length);
    check("cpuMs", d.queryTime?.cpuMs ?? null, t.cpuMs, (d.queryTime?.cpuMs ?? null) == t.cpuMs);
    check(
      "elapsedMs",
      d.queryTime?.elapsedMs ?? null,
      t.elapsedMs,
      (d.queryTime?.elapsedMs ?? null) == t.elapsedMs,
    );
    check(
      "grantedKb",
      d.memoryGrant?.grantedKb ?? null,
      t.grantedKb,
      (d.memoryGrant?.grantedKb ?? null) == t.grantedKb,
    );
    check(
      "missingIndexes",
      (d.missingIndexes ?? []).length,
      t.missingIndexes,
      (d.missingIndexes ?? []).length === t.missingIndexes,
    );

    // Every operator the digest reports must match the plan.
    // A cursor's plans reuse node ids, so match on id and operator name.
    const tops = t.operators;
    for (const o of d.topOperators) {
      const g =
        tops.find((x) => x.id === o.id && x.op === o.op) ??
        tops.find((x) => x.id === o.id);
      const w = `${where} node ${o.id} ${o.op}`;
      if (!g) {
        bad(short, w, "operator", "reported", "not in plan");
        continue;
      }
      operatorsChecked++;
      const f = (field, a, b, ok) => {
        fieldsChecked++;
        if (!ok) bad(short, w, field, a, b);
      };
      f("op", o.op, g.op, o.op === g.op);
      f("subtreeCost", o.subtreeCost, g.subtree, close(o.subtreeCost, g.subtree));
      f("ownCost", o.cost, g.own, close(o.cost, g.own, 1e-3) || Math.abs(o.cost - g.own) < 1e-5);
      f("estRows", o.estRows, g.estRows, close(o.estRows, g.estRows));
      if (g.threads) {
        f("actualRows", o.actualRows, g.actualRows, o.actualRows === g.actualRows);
        f("logicalReads", o.logicalReads ?? null, g.logicalReads, (o.logicalReads ?? null) == g.logicalReads);
        // A parallel operator runs once per thread: the digest counts it as executed once per logical
        // execution, so compare against executions per thread for those.
        const execs = g.parallel && g.threads > 1 ? Math.max(1, g.execs / (g.threads - 1)) : g.execs;
        f("execs", o.execs, g.execs, o.execs === g.execs || close(o.execs, execs, 0.01));
      }
      f("object", o.object ?? null, g.object, (o.object ?? null) === (g.object ?? null) || (o.object ?? "").startsWith(g.object ?? "\u0000"));
      f("lookup", !!o.lookup, !!g.lookup, !!o.lookup === !!g.lookup);
    }
    // The operators the models are told about must include the plan's costliest ones.
    const reported = new Set(d.topOperators.map((o) => o.id));
    const byOwn = [...t.operators].sort((a, b) => b.own - a.own)[0];
    if (byOwn && byOwn.own > 0)
      check("costliest operator reported", [...reported].join(","), byOwn.id, reported.has(byOwn.id));
    // Own elapsed time: row-mode operators report time including their children; batch mode reports its own.
    const ownMs = (o) =>
      o.elapsedMs == null
        ? 0
        : o.batch
          ? o.elapsedMs
          : o.elapsedMs -
            o.children
              .map((c) => t.operators.find((x) => x.id === c)?.elapsedMs ?? 0)
              .reduce((a, b) => a + b, 0);
    const slowest = [...t.operators].sort((a, b) => ownMs(b) - ownMs(a))[0];
    if (slowest && ownMs(slowest) >= 10)
      check("slowest operator reported", [...reported].join(","), slowest.id, reported.has(slowest.id));
    const byReads = [...t.operators].filter((o) => o.logicalReads).sort((a, b) => b.logicalReads - a.logicalReads)[0];
    if (byReads)
      check("most-read operator reported", [...reported].join(","), byReads.id, reported.has(byReads.id));
  }
}

console.log(
  `${files.length} plans, ${statementsChecked} statements, ${operatorsChecked} operators, ${fieldsChecked} fields checked: ${problems.length} disagreements`,
);
const byField = {};
for (const p of problems) byField[p.field] = (byField[p.field] ?? 0) + 1;
if (problems.length) console.log("by field:", byField);
for (const p of verbose ? problems : problems.slice(0, 40))
  console.log(`  ${p.file} | ${p.where} | ${p.field}: ours=${JSON.stringify(p.ours)} plan=${JSON.stringify(p.truth)}`);
process.exit(problems.length ? 1 : 0);

File: scripts/verify-parser.py
#!/usr/bin/env python3
"""Independent ground truth for ShowPlan XML, to check lib/plan-parser.mjs against.

Written against the ShowPlan schema with Python's ElementTree: a different language and XML parser that shares
no code with the app, so an agreement between the two is evidence, not a tautology. For every statement that has
a query plan it prints one JSON line with what a DBA would read off the plan:

  statement text and estimated cost, measured CPU/elapsed, memory grant, missing-index requests, and per
  operator: estimated subtree and own cost (subtree minus direct children), estimated rows and executions,
  actual rows, executions, logical reads and elapsed time summed/maxed over threads, the object it touches,
  and whether it is a lookup.

  python3 scripts/verify-parser.py plan.sqlplan [...]   (see scripts/verify-parser.mjs for the comparison)
"""
import json
import re
import sys
import xml.etree.ElementTree as ET

NS = "{http://schemas.microsoft.com/sqlserver/2004/07/showplan}"


def local(tag):
    return tag.split("}", 1)[-1]


def num(v):
    try:
        return float(v)
    except (TypeError, ValueError):
        return None


def read_xml(path):
    raw = open(path, "rb").read()
    for bom, enc in ((b"\xff\xfe", "utf-16-le"), (b"\xfe\xff", "utf-16-be"), (b"\xef\xbb\xbf", "utf-8-sig")):
        if raw.startswith(bom):
            return raw.decode(enc)
    if raw[:2] == b"<\x00":
        return raw.decode("utf-16-le")
    return raw.decode("utf-8")


def child_relops(relop):
    """RelOps directly below this one: the first RelOp on every path down, not deeper ones."""
    out = []

    def walk(el):
        for c in el:
            if local(c.tag) == "RelOp":
                out.append(c)
            else:
                walk(c)

    walk(relop)
    return out


def operator(relop):
    a = relop.attrib
    subtree = num(a.get("EstimatedTotalSubtreeCost")) or 0.0
    kids = child_relops(relop)
    own = subtree - sum(num(k.get("EstimatedTotalSubtreeCost")) or 0.0 for k in kids)
    threads = [
        t
        for rti in relop.findall(NS + "RunTimeInformation")
        for t in rti.findall(NS + "RunTimeCountersPerThread")
    ]
    def total(attr):
        vals = [num(t.get(attr)) for t in threads if t.get(attr) is not None]
        return sum(vals) if vals else None

    def most(attr):
        vals = [num(t.get(attr)) for t in threads if t.get(attr) is not None]
        return max(vals) if vals else None

    obj = None
    lookup = False
    for child in relop:
        if local(child.tag) in ("RelOp", "OutputList", "RunTimeInformation", "Warnings", "MemoryFractions"):
            continue
        if child.get("Lookup") in ("1", "true"):
            lookup = True
        o = child.find(NS + "Object")
        if o is not None:
            parts = [o.get(k) for k in ("Database", "Schema", "Table", "Index")]
            obj = ".".join(p.strip("[]") for p in parts if p)
            break
    return {
        "id": a.get("NodeId"),
        "op": a.get("PhysicalOp"),
        "subtree": subtree,
        "own": max(own, 0.0),
        "estRows": num(a.get("EstimateRows")),
        "estExecs": (num(a.get("EstimateRebinds")) or 0) + (num(a.get("EstimateRewinds")) or 0) + 1,
        "actualRows": total("ActualRows"),
        "execs": total("ActualExecutions"),
        "logicalReads": total("ActualLogicalReads"),
        "elapsedMs": most("ActualElapsedms"),
        "threads": len(threads),
        "parallel": a.get("Parallel") in ("1", "true"),
        "object": obj,
        "lookup": lookup or (a.get("PhysicalOp") in ("RID Lookup",)),
        "children": [k.get("NodeId") for k in kids],
        "batch": any(t.get("ActualExecutionMode") == "Batch" for t in threads),
    }


def statements(root):
    for stmt in root.iter():
        if local(stmt.tag) not in ("StmtSimple", "StmtCursor", "StmtCond", "StmtUseDb", "StmtReceive", "ExternalDistributedComputation"):
            continue
        if local(stmt.tag) == "StmtCond":
            continue  # its branches are statements of their own
        plans = [stmt.find(NS + "QueryPlan")]
        if plans[0] is None:
            # A cursor keeps one QueryPlan per operation (population, fetch) under CursorPlan.
            plans = list(stmt.iter(NS + "QueryPlan"))
            if not plans:
                continue
        plan = plans[0]
        ops = [operator(r) for p in plans for r in p.iter(NS + "RelOp")]
        qts = plan.find(NS + "QueryTimeStats")
        mg = plan.find(NS + "MemoryGrantInfo")
        missing = plan.findall(".//" + NS + "MissingIndex")
        yield {
            "statementId": stmt.get("StatementId"),
            "type": stmt.get("StatementType"),
            "sql": re.sub(r"\s+", " ", stmt.get("StatementText") or "").strip(),
            "cost": num(stmt.get("StatementSubTreeCost")),
            "cpuMs": num(qts.get("CpuTime")) if qts is not None else None,
            "elapsedMs": num(qts.get("ElapsedTime")) if qts is not None else None,
            "grantedKb": num(mg.get("GrantedMemory")) if mg is not None else None,
            "maxUsedKb": num(mg.get("MaxUsedMemory")) if mg is not None else None,
            "missingIndexes": len(missing),
            "operators": ops,
        }


def main():
    for path in sys.argv[1:]:
        try:
            text = re.sub(r"^\s*<\?xml[^>]*\?>", "", read_xml(path))  # the declared encoding no longer applies
            root = ET.fromstring(text)
        except ET.ParseError as e:
            print(json.dumps({"file": path, "error": str(e)}))
            continue
        for s in statements(root):
            s["file"] = path
            print(json.dumps(s))


if __name__ == "__main__":
    main()
